### PR TITLE
CLIPForImageClassification has been added

### PR DESCRIPTION
**Pull request by Andron00e**
# What does this PR do?

This PR adds a new model to the hub. Called CLIPForImage classification.

Details about implementation and pre-trained version on hub can be seen in my [repo](https://github.com/Andron00e/CLIPForImageClassification).

[My "New model" issue with describing of idea](https://github.com/huggingface/transformers/issues/27802)
[Model on hub link](https://huggingface.co/Andron00e/CLIPForImageClassification-v1)
[Code link](https://github.com/Andron00e/CLIPForImageClassification/blob/main/clip_for_classification/modeling_clipforimageclassification.py)

Tags:
vision models: @amyeroberts

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
-  Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/transformers/models/clip/modeling_clip.py [hunk 1]
```diff
@@ -1304,3 +1305,56 @@ def forward(
...
+    """
+    CLIP Model with a a linear layer on top of the CLIPOutput.image_embeds.
+    """,
+    CLIP_START_DOCSTRING,
+)
+class CLIPForImageClassification(CLIPPreTrainedModel):
+    """
+    Repo with custom implementation: https://github.com/Andron00e/CLIPForImageClassification
+    Pre-trained model on hub: https://huggingface.co/Andron00e/CLIPForImageClassification-v1
+    """
```
Comment (NielsRogge) [inline src/transformers/models/clip/modeling_clip.py:1320]: this can be removed

File: src/transformers/models/clip/modeling_clip.py [hunk 2]
```diff
@@ -1304,3 +1305,56 @@ def forward(
...
+    def __init__(self, config: CLIPConfig, num_labels: int):
+        super().__init__(config)
+        self.clip = CLIPModel.from_pretrained("openai/clip-vit-base-patch32")
```
Comment (NielsRogge) [inline src/transformers/models/clip/modeling_clip.py:1323]:
```suggestion
        self.vision_model = CLIPVisionTransformer(config)
```
In order to make sure people can load a CLIPForImageClassification model with pre-trained CLIP weights, we'll have to use the same variables.

File: src/transformers/models/clip/modeling_clip.py [hunk 3]
```diff
@@ -1304,3 +1305,56 @@ def forward(
...
+        for p in self.clip.parameters():
+            p.requires_grad=False
+        self.num_labels = num_labels
+        self.head = nn.Linear(self.config.projection_dim, num_labels)
+        self.processor = CLIPProcessor.from_pretrained("openai/clip-vit-base-patch32")
```
Comment (NielsRogge) [inline src/transformers/models/clip/modeling_clip.py:1328]:
```suggestion
        self.classifier = nn.Linear(self.config.projection_dim, config.num_labels)
```

File: src/transformers/models/clip/modeling_clip.py [hunk 4]
```diff
@@ -1304,3 +1305,56 @@ def forward(
...
+            position_ids=position_ids,
+            return_loss=return_loss,
+            output_attentions=output_attentions,
+            output_hidden_states=output_hidden_states,
+            return_dict=return_dict,
+        )
+
+        logits = self.head(clip_outputs.image_embeds)
+
+        loss = None
```
Comment (NielsRogge) [inline src/transformers/models/clip/modeling_clip.py:1355]: here, the same logic as here needs to be defined: https://github.com/huggingface/transformers/blob/4d4febb7aa3f49c29d7f4fc29bbc1760edd583b6/src/transformers/models/dinov2/modeling_dinov2.py#L721-L757

**Review by NielsRogge — COMMENTED** [review]
Overall it's a good idea to add a CLIPForImageClassification, but one would need to make sure weights can be loaded using `from_pretrained`.

File: src/transformers/models/clip/modeling_clip.py [hunk 5]
```diff
@@ -1304,3 +1305,56 @@ def forward(
...
+        self.clip = CLIPModel.from_pretrained("openai/clip-vit-base-patch32")
+        for p in self.clip.parameters():
+            p.requires_grad=False
+        self.num_labels = num_labels
+        self.head = nn.Linear(self.config.projection_dim, num_labels)
+        self.processor = CLIPProcessor.from_pretrained("openai/clip-vit-base-patch32")
+
+    def forward(
+        self,
+        input_ids: Optional[torch.LongTensor] = None,
```
[the remark this answers is not in the document]
Comment (Andron00e) [reply src/transformers/models/clip/modeling_clip.py:1330]: You are right, there is no need in "input_ids" in this case. 
But it is a key element when we use clip. for my further enhancement I am going to output not "clip_output.image_embeddings" but "logits_per_image" and here we need those "input_ids".

**Review by Andron00e — COMMENTED** [review]
In my opinion it would be better to save this structure of forward pass for further enhancement.

**Pull request closed without merging**